consensus/parlia/fusionfork.go

```diff
@@ -0,0 +1,226 @@
+	// not just its contents.
+	*h = append(*h, x.(ValidatorItem))
+}
+
+func (h *ValidatorHeap) Pop() interface{} {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+func (p *Parlia) updateEligibleValidators(state *state.StateDB, header *types.Header, chain core.ChainContext,
+	txs *[]*types.Transaction, receipts *[]*types.Receipt, receivedTxs *[]*types.Transaction, usedGas *uint64, mining bool,
+) error {
+	// 1. get all validators and its voting power
+	method := "getValidatorWithVotingPower"
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel() // cancel when we are finished consuming integers
+
+	blockNr := rpc.BlockNumberOrHashWithHash(header.ParentHash, false)
+	toAddress := common.HexToAddress(systemcontracts.StakeHubContract)
+	gas := (hexutil.Uint64)(uint64(math.MaxUint64 / 2))
+
+	var allValidators []common.Address
+	var allVotingPowers []*big.Int
+	for {
+		data, err := p.stakeHubABI.Pack(method, big.NewInt(int64(len(allValidators))), big.NewInt(100))
+		if err != nil {
+			log.Error("Unable to pack tx for getValidatorWithVotingPower", "error", err)
+			return err
+		}
+		msgData := (hexutil.Bytes)(data)
+
+		result, err := p.ethAPI.Call(ctx, ethapi.TransactionArgs{
+			Gas:  &gas,
+			To:   &toAddress,
+			Data: &msgData,
+		}, blockNr, nil, nil)
+		if err != nil {
+			return err
+		}
+
+		var validators []common.Address
+		var votingPowers []*big.Int
+		var totalLength *big.Int
+		if err := p.stakeHubABI.UnpackIntoInterface(&[]interface{}{&validators, &votingPowers, &totalLength}, method, result); err != nil {
+			return err
+		}
+
+		allValidators = append(allValidators, validators...)
+		allVotingPowers = append(allVotingPowers, votingPowers...)
+		if totalLength.Cmp(big.NewInt(int64(len(allValidators)))) == 0 {
+			break
+		}
+	}
+
+	// 2. sort by voting power
+	method = "maxElectedValidators"
+	data, err := p.stakeHubABI.Pack(method)
+	if err != nil {
+		log.Error("Unable to pack tx for maxElectedValidators", "error", err)
+		return err
+	}
+	msgData := (hexutil.Bytes)(data)
+	result, err := p.ethAPI.Call(ctx, ethapi.TransactionArgs{
+		Gas:  &gas,
+		To:   &toAddress,
+		Data: &msgData,
+	}, blockNr, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	var maxElectedValidators *big.Int
+	if err := p.stakeHubABI.UnpackIntoInterface(&maxElectedValidators, method, result); err != nil {
+		return err
+	}
+
+	var validatorHeap ValidatorHeap
+	if maxElectedValidators.Cmp(big.NewInt(int64(len(allValidators)))) == 1 {
+		for i := 0; i < len(allValidators); i++ {
+			if allVotingPowers[i].Cmp(big.NewInt(0)) == 1 {
+				validatorHeap = append(validatorHeap, ValidatorItem{
+					address:     allValidators[i],
+					votingPower: allVotingPowers[i],
+				})
+			}
+		}
+		sort.SliceStable(validatorHeap, validatorHeap.Less)
+	} else {
+		i := 0
+		for len(validatorHeap) < int(maxElectedValidators.Int64()) && i < len(allValidators) {
```

Comment:
==> 
```
for  i:=0; len(validatorHeap) < int(maxElectedValidators.Int64()) && i < len(allValidators); i++ {
  if allVotingPowers[i].Cmp(big.NewInt(0)) == 1 {
				validatorHeap = append(validatorHeap, ValidatorItem{
					address:     allValidators[i],
					votingPower: allVotingPowers[i],
				})
			}
}
```

Reply:
I think there is a bug here:  the validator with more voting power may be at the end of the slice of `allValidators`, and thus we are not able to put it into `validatorHeap`